### DISCUSSION TITLE
Batch Process in Mirth

### DISCUSSION THREAD

**Discussion by Govin-Rai09** [nextgenhealthcare/connect#5328] [Q&A]
I am performing some functions on a file through a JAR in Mirth. Upon completion I have a function present in JAR that should be able to delete the file.  
But I am not able to delete the file and get the
                   "The process could not be completed because the file is being used by another process"     error.
If I uncheck the Process Batch the function works.

Is the batch process making the file open so I cannot delete it?
FYI we have to perform the delete operation through JAR only.

**Comment by mjtrac** [comment]
Is the last item in the batch coming through properly?  Can you set the response to respond after the last message?  The first thing that comes to mind is that perhaps Mirth's parser doesn't see a proper ending to the batch.  If the read times out, you might call the jar file deletion function from your channel's postprocessor script.  (This is just speculation, I haven't tested.)

Mirth has a finally clause closing the file and connection at approx line 400 in function processFile, starting around line 350 here: https://github.com/nextgenhealthcare/connect/blob/development/server/src/com/mirth/connect/connectors/file/FileReceiver.java

**Reply by Govin-Rai09** [reply]
The function I am trying is to move a file from one location to another and delete it from the source. The file is getting moved but since I am not able to delete the file in the source path , the same file is getting moved again and again.

**Reply by mjtrac** [reply]
If I'm understanding you correctly, you're not using Mirth's after-processing capability to move the file, but a function in a jar file you've loaded as a resource.  At least AFAIK, an operating system move is going to remove the file from the source location (actually fiddling with directory entries) and I'd assume Java does the same.  If you are calling a function that moves the file, and calling it from anywhere but the Postprocessor script (accessible from the Scripts tab) I'd try moving the call to the Postprocessor script and seeing if that resolves the problem.  You'd only want it to be invoked once after the last message has been processed from the batch, and you might need to test for that condition in the postprocessor script.  There should be a map variable that will tell you when the Postprocessor is getting called for the final message of the batch. It might also help to call sleep or an equivalent in the script, just prior to calling your function.

I sort of hesitate to add this, but you might also use Mirth's own move operation to move the file to a "processed" subfolder you create, and then change the arguments to your jar file function to pick it up from there.  If Mirth is able to move the file, you ought to be able to move it as well.
